multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,275 @@
+}
+
+// onPodCreateOrUpdate updates the Pod and label identity mapping in the cache, and
+// updates label identity ResourceExport if necessary (the Pod creation/update
+// event causes a new label identity to appear in the cluster or a label identity
+// to no longer being present in the cluster).
+func (r *LabelIdentityReconciler) onPodCreateOrUpdate(ctx context.Context, podKey, normalizedLabel string) error {
+	labelToAdd, labelToDelete := r.getLabelToUpdate(podKey, normalizedLabel)
+	if labelToDelete != "" {
+		if err := r.deleteLabelIdentityResourceExport(ctx, labelToDelete); err != nil {
+			return err
+		}
+		r.updateLabelCache(podKey, "", labelToDelete)
+	}
+	if labelToAdd != "" {
+		if err := r.addLabelIdentityResourceExport(ctx, labelToAdd); err != nil {
+			return err
+		}
+		r.updateLabelCache(podKey, labelToAdd, "")
+	}
+	return nil
+}
+
+// getLabelToUpdate gets all label identities to be added and to be deleted due to Pod
+// update event. It is protected with labelMutex so that concurrent Pod update events
+// will not interfere with the label update calculations.
+func (r *LabelIdentityReconciler) getLabelToUpdate(podKey, normalizedLabel string) (labelToAdd string, labelToDelete string) {
+	r.labelMutex.RLock()
+	defer r.labelMutex.RUnlock()
+
+	originalLabel, isCached := r.podLabelCache[podKey]
+	if isCached && originalLabel != normalizedLabel {
+		labelToDelete = originalLabel
+	}
+	if normalizedLabel != "" {
+		labelToAdd = normalizedLabel
+	}
+	return
+}
+
+// updateLabelCache updates podLabelCache and labelToPodsCache once the ResourceExport
+// create/update/delete operation is successful.
+func (r *LabelIdentityReconciler) updateLabelCache(podKey, updatedLabel, deletedLabel string) {
+	r.labelMutex.Lock()
+	defer r.labelMutex.Unlock()
+
+	if updatedLabel != "" {
+		r.podLabelCache[podKey] = updatedLabel
+		if _, ok := r.labelToPodsCache[updatedLabel]; !ok {
+			r.labelToPodsCache[updatedLabel] = sets.NewString(podKey)
+		}
+		r.labelToPodsCache[updatedLabel].Insert(podKey)
+	}
+	if deletedLabel != "" {
+		delete(r.podLabelCache, podKey)
+		if podNames, ok := r.labelToPodsCache[deletedLabel]; ok {
+			podNames.Delete(podKey)
+			if len(podNames) == 0 {
+				klog.V(2).InfoS("Label no longer exists in the cluster", "label", deletedLabel)
+				delete(r.labelToPodsCache, deletedLabel)
+			}
+		}
+	}
+}
+
+func (r *LabelIdentityReconciler) addLabelIdentityResourceExport(ctx context.Context, normalizedLabel string) error {
+	existingResExport := &mcsv1alpha1.ResourceExport{}
+	resNamespaced := types.NamespacedName{
+		Name:      getResourceExportNameForLabelIdentity(r.localClusterID, normalizedLabel),
+		Namespace: r.remoteCommonArea.GetNamespace(),
+	}
+	err := r.remoteCommonArea.Get(ctx, resNamespaced, existingResExport)
+	if err != nil && !apierrors.IsNotFound(err) {
+		klog.ErrorS(err, "Failed to get ResourceExport in remote cluster", "resourceexport", resNamespaced)
+		return err
+	}
+	if apierrors.IsNotFound(err) {
```

Comment:
The client cache may be stale (e.g. another event is deleting the export), this seems not safe.

Please think about the race condition with deletion too.

Reply:
Yes added back the ResourceExport mutex

Reply:
I feel the mutex does not help, as it does not cover the client cache update, which is async in my understanding. Why we do not check labelToPodsCache instead?

Reply:
The `labelToPodsCache` also has a mutex around it . I reasoned a little bit deeper, and figured that maybe the `onPodDelete` and `onPodCreateOrUpdate` operations all need to be atomic. I tried to do something “clever” where concurrent access to `labelToPodsCache` and concurrent updates to ResourceExport are protected individually while the two operations themselves are asynchronous. This was mainly to refrain costly operations (K8s API calls) from blocking cache reads etc. However, I found out there are cases that this approach breaks, no matter we update the cache before or after successful ResourceExport operation or not:

1. If we only update the `labelToPodsCache` after the API calls are successful, the calculations for label add will not be accurate if there’s an in-flight deletion of the same label. I.e. one worker tries to delete the label, releases `labelMutex`, and calls ResourceExport delete. At that very moment, another worker is handling add of this particular label for another Pod. The ResourceExport add event will be missed since at that specific moment the label already exists in cache and is not yet deleted.
2. If we update the `labelToPodsCache` immediately and revert cache in case of API failure, other problem can occur. Consider two different workers handling add for the same label at roughly the exact same time. The first worker (reconciling pod1) acquires the `labelMutex` will add the label to cache and initiate ResourceExport create with K8s API. The second worker (reconciling pod2) will only add the pod2’s name to the cached label, since it is a known label already. Now if the API call fails, the first reconcile will be retried, and the pod1 to label mapping will be reverted. However, by the time the reconcile is retried, the label is already in cache since pod2 is referring to it. The reconciler will then skip ResourceExport creation, which is a missed event.

Hence, I think we will have to use the same mutex around the entire “check cache, do API call if necessary, update cache” operation, even if it makes it harder for reconciler workers to do parallel work. cc @GraysonWu for awareness

Reply:
My earlier comment is orthogonal to the problem you described. I meant remoteCommonArea.Get() should read from the K8s client cache, which may not be up to date. And mutex may not solve the problem, as the client cache update may be async (not right after update or deletion call returns)? So, I am suggesting instead of calling Get(), could we check labelToPodsCache to know if creation is needed.

Then for the problem you described, I think you can still update labelToPodsCache after a successful creation/deletion.  Even if race occurs (which should be rare), it will just end up with an unnecessary creation/deletion call and you can ignore the exists/not-found error with no harm.

Reply:
I think I got your point. We already use `labelToPodsCache` to check whether a ResourceExport creation is needed. And cache check/update is mutex protected. I think the reason why it was confusing is because in this specific function, it calls `remoteCommonArea.Get()` to validate if such ResourceExport is indeed not there, which is unnecessary. Updated the code to simply create ResourceExport when needed (indicated by the `labelToPodsCache`) and ignore already exist errors.

Reply:
I think I got your point. We already use `labelToPodsCache` to check whether a ResourceExport creation is needed. And cache check/update is mutex protected. I think the reason why it was confusing is because in this specific function, it calls `remoteCommonArea.Get()` to validate if such ResourceExport is indeed not there, which is unnecessary. Updated the code to simply create ResourceExport when needed (indicated by the `labelToPodsCache`) and ignore already exist errors.